[Debugger] Cocoa port of ObjectValueTreeView

jstedfast opened the pull request:
This is a rewrite of the current Gtk UI for the Locals Pad, Watch Pad, Debugger Tooltips and Pinned Watches using Xamarin.Mac.

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/957702/
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/888712/
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/901729/
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/782188/
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/754947/

---

There are a number of issues still remaining in this PR and this comment is here to try and track them (thus it will be updated as new issues are found or issues become resolved).

These are the known issues remaining:

- [x] In local builds of this PR, the Locals & Watch pads are scrollable no matter how many nodes in the list are expanded, but in the CI builds, this is not the case. When using a CI build, expanding nodes seems to trick the `NSScrollView` containing the `MacObjectValueTreeView` (a subclass of `NSOutlineView`) into thinking that it has far more visible real estate than it actually has and therefor it does not allow you to scroll. This is just a theory, but it makes sense, I think? Perhaps an issue with GtkNSHostView?
- [x] Adding a new expression in the Watch Pad only semi-sorta works. For some reason, when adding an expression that would require an expander (e.g. if you add the expression `buffer` which is defined as `new byte[123]`), the expression string is NOT displayed in the new row that gets added (but the expander, icon, and preview eyeball icon thingy *do* get shown). Based on what I can figure out so far, the expression string *is* being set on the `TextField`, but it's not rendering... is it just oddly sized to be 0px wide or something??
- [x] In the Locals and Watch Pads, sometimes when an expression that has a visualizer icon in the Value column is pushed off-screen (due to an expression above it being expanded, for example), when you collapse said expanded node allowing the other expression back into the viewport, the visualizer icon is gone. Once it is gone, it seems to never come back. 
![Screen Recording 2019-10-03 at 03 02 PM](https://user-images.githubusercontent.com/338984/66156391-91f1d400-e5ef-11e9-8b0b-2e3f4c2d23ea.gif)
- [x] While @sgmunn was playing around with the Locals Pad, he discovered that expanding the "Non-Public Members" row, it would insert X-number of completely blank rows into the treeview (presumably where X matches the number of non-public members). See the screenshot below: <img width="710" alt="Screen Shot 2019-10-03 at 11 18 02 AM" src="https://user-images.githubusercontent.com/338984/66142899-45e56600-e5d4-11e9-8dbf-e2f7998aa148.png">
- [x] Unable to resize columns in Locals/Watch pads: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/993219/
- [x] Sometimes text in the tooltip is not displayed until you click inside the tooltip: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/993245/
- [x] Rows do not stay expanded when stepping like the Gtk one does
- [x] "Add new expression" placeholder text is not very visible when the row is selected: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/993212/
- [x] context menus are missing for locals and watch windows
- [x] Greg has also been able to confuse the Locals Pad into display the preview window in the wrong location: <img width="512" alt="Screen Shot 2019-10-03 at 11 18 20 AM" src="https://user-images.githubusercontent.com/338984/66142969-69a8ac00-e5d4-11e9-980a-6e6e63c96079.png"> When Greg got encountered this bug, he had first resized the Locals Pad to be much larger and then back to a smaller size, so this *may* be required step to reproduce this bug (unconfirmed). @mrward also seems to have encountered this issue (or something similar) with https://devdiv.visualstudio.com/DevDiv/_workitems/edit/993216/
- [x] rendering in tooltips, tree reloading issues
![Screen Recording 2019-10-07 at 09 56 AM](https://user-images.githubusercontent.com/723721/66318355-265a8000-e8e9-11e9-8165-002ba8eedd69.gif)
- [x] column headers are transparent, looks odd with the rest of the IDE.
- [x] Pop-up window does not close when the parent pad window is clicked. Can be open when trying to add a new watch expression.
<img width="331" alt="TooltipDoesNotClose" src="https://user-images.githubusercontent.com/372361/66488103-5189ce80-eaa5-11e9-9aa5-9326e0070c25.png">

- [ ] No code completion for watches (regression)
- [ ] Need to use the `~sel` icons when the row is selected (see @vancura 's comment above) - tracking bug: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/993725
- [ ] Unable to reposition the pinned watch window within the editor: (regression, kinda) https://devdiv.visualstudio.com/DevDiv/_workitems/edit/993254/
- [ ] Pinned watch windows are not scrollable (regression, sort of).
- [ ] Clicking the refresh icon in a pinned tooltip does not update the value. (not a regression)
- [ ] Backspace key does not remove selected watch expression. Have to select all the text and remove it to remove the watch expression.
- [x] does the locals and watch pads support custom fonts? there is a todo for that
- [x] (as designed) pinned watch is hidden behind the bottom pad when there's overlap (Note: this is a limitation of text editor adornments - they can't sandwich the rest of the IDE)
![image](https://user-images.githubusercontent.com/34032260/66519728-d846a780-eab5-11e9-92d9-4acdeab06aac.png)

- [ ] pinned watches are removed when scrolled out of view in the editor, which leads to a native crash on stopping he debugger it seems. (need to grab logs for that).
- [ ] pinned watches don't stay expanded when stepping. they collapse but the size stays the same.
- [ ] changing pad general text does not update in the locals or watch pads

jstedfast commented:
Screenshots:

This is the Xamarin.Mac implementation of the Locals Pad:
<img width="1680" alt="mac-locals-pad" src="https://user-images.githubusercontent.com/338984/63967407-594c6100-ca6b-11e9-8d43-89ff952d0ea8.png">

This is the Xamarin.Mac implementation of the debugger tooltips:
<img width="1680" alt="mac-debugger-tooltip" src="https://user-images.githubusercontent.com/338984/63967458-75e89900-ca6b-11e9-8cfe-e7bd45115b6d.png">

Expanded tooltip:
<img width="1680" alt="mac-debugger-tooltip-expanded" src="https://user-images.githubusercontent.com/338984/63967468-7e40d400-ca6b-11e9-97ed-1826a593d969.png">

This is the Xamarin.Mac implementation of the pinned watch text editor adornment (basically just a pinned tooltip):
<img width="1680" alt="mac-pinned-watch" src="https://user-images.githubusercontent.com/338984/63967501-9284d100-ca6b-11e9-8d82-7a9981abdc09.png">

Same, but expanded:
<img width="1680" alt="mac-pinned-watch-expanded" src="https://user-images.githubusercontent.com/338984/63967524-a03a5680-ca6b-11e9-803d-2d1e7b1bcbc6.png">

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/PinnedWatches/PinnedWatchProvider.cs:
```diff
@@ -0,0 +1,54 @@
+		public void TextViewCreated (ICocoaTextView textView)
+		{
+			var manager = new PinnedWatchAdornmentManager (cocoaViewFactory, textView);
+			textView.Closed += (s, e) => manager.Dispose ();
```
For consistency, I think we should probably follow the pattern used by most VS-editor stuff:
https://github.com/mono/monodevelop/blob/f9839cdd26fd7710315b7ad7fcb7bc7b585223c9/main/src/addins/MonoDevelop.SourceEditor2/VSEditor/Language/Impl/Intellisense/IntellisenseSessionStackMapService.cs#L36-L63
https://github.com/mono/monodevelop/blob/305bddc1e9dcd90caf25f8bed1a042efc317528b/main/src/addins/MonoDevelop.SourceEditor2/VSEditor/Language/Impl/Intellisense/MultiSessionIntellisensePresenterProvider.cs#L69-L99

Add the manager to a the text view's properties and via the provider and the provider is responsible for removing it and disposing it.

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/PinnedWatches/PinnedWatchView.cs:
```diff
@@ -0,0 +1,195 @@
+			if (disposing && !disposed) {
+				DebuggingService.ResumedEvent -= OnDebuggerResumed;
+				DebuggingService.PausedEvent -= OnDebuggerPaused;
+				treeView.Resized += OnTreeViewResized;
```
```suggestion
				treeView.Resized -= OnTreeViewResized;
```

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/PinnedWatches/PinnedWatchAdornmentManager.cs:
```diff
@@ -0,0 +1,189 @@
+			this.cocoaViewFactory = cocoaViewFactory;
+			this.textView = textView;
+
+			this.textView.LayoutChanged += OnTextViewLayoutChanged;
```
this isn't unsubscribed and the method is empty?

jstedfast replied:
Thanks, I'll add it.

Therzok commented on main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor/DebugValueTooltipProvider.cs:
```diff
@@ -24,7 +24,9 @@
-//
+
+// Note: This API is only used by the old (Gtk) TextEditor.
+// The new TextEditor uses MonoDevelop.Debugger.VSTextView.QuickInfo.DebuggerQuickInfoSourceProvider
```
Maybe obsolete the class?

jstedfast replied:
sure.

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/PinnedWatches/PinnedWatchAdornmentManager.cs:
```diff
@@ -0,0 +1,189 @@
+
+namespace MonoDevelop.Debugger.VSTextView.PinnedWatches
+{
+	public class PinnedWatchAdornmentManager : IDisposable
```
Does this need to be public?

jstedfast replied:
No. I've made it internal now.

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/PinnedWatches/PinnedWatchAdornmentManager.cs:
```diff
@@ -0,0 +1,189 @@
+			debugging = false;
+		}
+
+		void OnTextViewLayoutChanged (object sender, TextViewLayoutChangedEventArgs e)
```
Empty method.

jstedfast replied:
The code is actually only commented out because I think there's a bug in the TextEditor that emits these scroll events when scrolling and not editing the content.

It's commented out momentarily because I need to be able to test scrolling out.

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/PinnedWatches/PinnedWatchView.cs:
```diff
@@ -0,0 +1,195 @@
+
+namespace MonoDevelop.Debugger.VSTextView.PinnedWatches
+{
+	public class PinnedWatchView : NSScrollView
```
Same here, does this need to be public?

Therzok replied:
```suggestion
	sealed class PinnedWatchView : NSScrollView
```

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/PinnedWatches/PinnedWatchView.cs:
```diff
@@ -0,0 +1,195 @@
+
+			// Find our parent CocoaTextViewControl
+			var textView = materialView;
+			while (textView != null && textView.GetType ().Name != CocoaTextViewControl)
```
Can we just look for `ITextView3Container`? cc @sandyarmstrong

sandyarmstrong replied:
Well you could, yes, but unfortunately we haven't moved `ITextView3Container` to `vs-editor-api` yet so I think you won't be able to resolve the type here.

I'm also not sure that the `ITextView3Container` type is still used. If it is, we should move it. If not, we should remove it.

I think the current check is fine.

sandyarmstrong replied:
(I checked and the interface _is_ still used. But as it's still not public, I still think this line of code is fine for now.)

Therzok replied:
Well, last I checked it was public: https://github.com/xamarin/vs-editor-core/blob/master/src/Editor/Text/Impl/CocoaView/TextUI/Editor/ITextView3Container.cs#L11

sandyarmstrong replied:
Sorry, I misspoke. It's not _publicly available_. MonoDevelop does not have a reference that allows it to access the type.

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/QuickInfo/DebuggerQuickInfoSource.cs:
```diff
@@ -119,36 +125,54 @@ private async Task EvaluateAndShowTooltipAsync (IAsyncQuickInfoSession session,
+			location.EndColumn = column;
+			location.EndLine = line;
+
+			window = new MacDebuggerTooltipWindow (location, DebuggingService.CurrentFrame, val, null);
```
Nit, named parameter here to know what null is being passed to.

Therzok replied:
```suggestion
			window = new MacDebuggerTooltipWindow (location, DebuggingService.CurrentFrame, val, watch: null);
```

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/QuickInfo/DebuggerQuickInfoSource.cs:
```diff
@@ -188,19 +214,11 @@ private void LayoutChanged (object sender, TextViewLayoutChangedEventArgs e)
 			if (window != null) {
-				window.Destroy ();
-				window.LeaveNotifyEvent -= LeaveNotifyEvent;
+				window.Close ();
```
Nit: also Dispose it to clear the managed ref.

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/QuickInfo/MacDebuggerTooltipWindow.cs:
```diff
@@ -0,0 +1,152 @@
+
+namespace MonoDevelop.Debugger
+{
+	class MacDebuggerTooltipWindow : NSPopover
```
Nit:
```suggestion
	class MacDebuggerTooltipWindow : NSPopover
```

jstedfast replied:
What's the nit? I'm not seeing a difference?

Therzok replied:
Not sure how github let me do that. It was to seal the class. (It might have an impact on the xammac registrar)

Therzok replied:
Ah, no, seems like it'll only be useful when we actually use the managed linker.

jstedfast replied:
Ah! Fixed.

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/QuickInfo/MacDebuggerTooltipWindow.cs:
```diff
@@ -0,0 +1,152 @@
+			heightConstraint = scrollView.HeightAnchor.ConstraintEqualToConstant (treeView.Frame.Size.Height);
+			heightConstraint.Active = true;
+
+			treeView.Resized += OnTreeViewResized;
```
Move this with the others?

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/QuickInfo/MacDebuggerTooltipWindow.cs:
```diff
@@ -0,0 +1,152 @@
+				Frame = treeView.Frame
+			};
+
+			ContentViewController = new NSViewController (null, null) {
```
Can we just use the parameterless constructor?

jstedfast replied:
Apparently, yes. I just saw existing code somewhere using (null, null)

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/QuickInfo/MacDebuggerTooltipWindow.cs:
```diff
@@ -0,0 +1,152 @@
+
+		void OnTreeViewResized (object sender, EventArgs e)
+		{
+			var maxHeight = (float) GetMaxHeight (treeView.Window);
```
We can keep using nfloat. `System.NMath.Min` should do the trick.

jstedfast replied:
Thanks, I didn't know that existed!

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/QuickInfo/MacDebuggerTooltipWindow.cs:
```diff
@@ -0,0 +1,152 @@
+		readonly ObjectValueTreeViewController controller;
+		readonly NSLayoutConstraint heightConstraint;
+		readonly MacObjectValueTreeView treeView;
+		readonly NSScrollView scrollView;
```
This isn't used anywhere outside the constructor.

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/LocalsPad.cs:
```diff
@@ -53,21 +53,21 @@ void ReloadValues ()
 				controller.ClearValues ();
 				controller.AddValues (locals);
 
-				//var xx = new System.Collections.Generic.List<ObjectValueNode> ();
```
Can we put this in another method so we only comment/uncomment a method call line?
```csharp
void AddFakeNodes()
{
    if (!FeatureSwitchService.IsFeatureEnabled ("DEBUGGER_FAKE_NODES"))
        return;
...
}
```

Or using compilation macros at the top of the file:
```csharp
#define FAKE_NODES

[Conditional ("FAKE_NODES")]
void AddFakeNodes(...)
{
    ...
}
```

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/AddNewExpressionObjectValueNode.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace MonoDevelop.Debugger
+{
+	public class AddNewExpressionObjectValueNode : ObjectValueNode
```
Should this be public?
```suggestion
	public sealed class AddNewExpressionObjectValueNode : ObjectValueNode
```

Otherwise:
```suggestion
	sealed class AddNewExpressionObjectValueNode : ObjectValueNode
```

jstedfast replied:
Doesn't need to be public.

netonjm commented:
Some visual fixes:
![image](https://user-images.githubusercontent.com/1587480/64042707-cd1d6500-cb62-11e9-927d-171be3bae66f.png)

![image](https://user-images.githubusercontent.com/1587480/64042877-2daca200-cb63-11e9-93bf-512c32a87b7f.png)

vancura commented:
@netonjm Had really good points, I currently don't have anything to add. @jstedfast could you please upload new screenshots? Thanks!

jstedfast commented:
@vancura I haven't fixed any of the UI issues pointed out by @netonjm yet, but will give you a poke when I do. Thanks!

jstedfast commented:
@vancura Here are the new screenshots:

<img width="793" alt="LocalsPad" src="https://user-images.githubusercontent.com/338984/64447242-e2d1e380-d0a8-11e9-9bef-1474a36c5870.png">
<img width="793" alt="Tooltip" src="https://user-images.githubusercontent.com/338984/64447244-e5ccd400-d0a8-11e9-94a1-4490f196b310.png">
<img width="793" alt="PinnedWatch" src="https://user-images.githubusercontent.com/338984/64447249-e82f2e00-d0a8-11e9-95b7-dd51bc0fa5ff.png">

vancura commented:
This looks great, I'll just have a couple of details, as usual :)

- [ ] Is it possible to vertically center labels inside the table? See this example, left: your version, right: my tweak. I just moved the labels `1pt` up. That applies to the popover too.

![image](https://user-images.githubusercontent.com/4982/64526324-46efe400-d303-11e9-9052-d3c77bbc4a40.png)

---

- [ ] Please also vertically center the label inside the gray bubble, move it `1pt` up:

![image](https://user-images.githubusercontent.com/4982/64526575-fb8a0580-d303-11e9-90a6-145f91020bb9.png)

---

- [ ] The popover has wrong sizes on the edges. The table row height is `19pt`, while the top row is trimmed from the top (resulting in `18pt`) and the last visible row is trimmed from the bottom (resulting in `18pt` as well). Please make the popover height `2pt` more, and make sure these rows don't get trimmed.

![image](https://user-images.githubusercontent.com/4982/64526856-af8b9080-d304-11e9-93eb-453571d3d536.png)

Thanks!

vancura commented:
Eek. Ignore that one then, AppKit has it's own opinions again :)

Could you please fix the rest and send me some screenshots? Thanks!

jstedfast commented:
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/934130/
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/957702/
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/888712/
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/901729/

jstedfast commented:
I'm not sure the other UI issues can be fixed either, because the only way to fix them that I can find is to hardcode pixel offsets which will blow up in our face if/when the user changes font sizes.

jstedfast commented:
<img width="717" alt="fixed-text-alignment" src="https://user-images.githubusercontent.com/338984/65172073-5b765f80-da1a-11e9-862c-2f4dc0f35e19.png">

vancura commented:
This looks amazing! Thanks a lot for the changes, @jstedfast!

DavidKarlas commented:
I found 2 issues:
1) When number of lines in Locals pad goes beyond size of Locals pad scrolling doesn't work:
<img width="565" alt="image" src="https://user-images.githubusercontent.com/774791/65367459-4c69fa00-dc32-11e9-8b86-1147739b4a8b.png">
2) Resizing of tooltip doesn't seem to ideal:
In this case its too big:
<img width="298" alt="image" src="https://user-images.githubusercontent.com/774791/65367466-715e6d00-dc32-11e9-8c63-916dcd7b6e30.png">
In this case its too small:
<img width="273" alt="image" src="https://user-images.githubusercontent.com/774791/65367472-8509d380-dc32-11e9-89b5-a4dee894f63d.png">
Whole tooltip needs to be resizing to bigger size as long as possible, if it gets too big for screen it should use scrolling, for example, name of properties must always be fully visible(5x`Total` is not very useful, also `Non-` is weird... Also there is some extra space after Pin icon that seems unnessesary.

I want to add responsiveness difference to old debugger tooltip is astonishing. Great job so far!

jstedfast commented:
> Also there is some extra space after Pin icon that seems unnessesary.

that's been fixed now. The extra space is needed for pinned watches (the extra space ends up being for a refresh button).

DavidKarlas commented:
What about 3 other problems I mentioned? :)

iainx commented:
Can I check the accessibility for this before it's merged please?

iainx commented:
* Locals pad:
   - Not accessible, need to investigate why
* Debugger tooltip in editor:
   - Pin button has no `AccessibilityTitle` set, so VoiceOver just reads "Button"
   - Pin button isn't shown except on mouseover. If accessibility is currently being used, then we should always show the button. Use `IdeServices.DesktopService.AccessibilityInUse` to check for this
   - Not sure how to access Debugger tooltip when it is pinned, I suspect this will require the editor accessibility work to make adornments accessible

jstedfast commented:
@iainx the Pin button has an accessibility title now and will always show up if accessibility is in use. Thanks for the tips on how to do that.

jstedfast commented:
@DavidKarlas the LocalsPad scrolls just fine for me when the list grows longer than the height of the pad...

iainx commented:
@jstedfast awesome, thanks. I'm looking into the Locals pad issue.

I hit this crash while playing with it. I started a Xammac project from the default template, and paused it in the debugger. There was a single row in the Locals pad with the name "main" and the disclosure arrow was pointing downward, showing that the row was expanded, but there were no other rows. I clicked the arrow to close the row and this happened

```System.AggregateException: One or more errors occurred. (Value cannot be null.
Parameter name: str) ---> System.ArgumentNullException: Value cannot be null.
Parameter name: str
  at Foundation.NSAttributedString..ctor (System.String str, Foundation.NSDictionary attributes) [0x00098] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.16.1.24/src/Xamarin.Mac/Foundation/NSAttributedString.g.cs:298 
  at Foundation.NSAttributedString..ctor (System.String str, AppKit.NSFont font, AppKit.NSColor foregroundColor, AppKit.NSColor backgroundColor, AppKit.NSColor strokeColor, AppKit.NSColor underlineColor, AppKit.NSColor strikethroughColor, AppKit.NSUnderlineStyle underlineStyle, AppKit.NSUnderlineStyle strikethroughStyle, AppKit.NSParagraphStyle paragraphStyle, System.Single strokeWidth, AppKit.NSShadow shadow, Foundation.NSUrl link, System.Boolean superscript, AppKit.NSTextAttachment attachment, Foundation.NSLigatureType ligature, System.Single baselineOffset, System.Single kerningAdjustment, System.Single obliqueness, System.Single expansion, AppKit.NSCursor cursor, System.String toolTip, System.Int32 characterShape, AppKit.NSGlyphInfo glyphInfo, Foundation.NSArray writingDirection, System.Boolean markedClauseSegment, AppKit.NSTextLayoutOrientation verticalGlyphForm, AppKit.NSTextAlternatives textAlternatives, AppKit.NSSpellingState spellingState) [0x00000] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.16.1.24/src/Xamarin.Mac/Foundation/NSAttributedString.mac.cs:54 
  at MonoDevelop.Debugger.MacDebuggerObjectCellViewBase.GetAttributedString (System.String text) [0x0000e] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacDebuggerObjectCellViewBase.cs:130 
  at MonoDevelop.Debugger.MacDebuggerObjectTypeView.UpdateContents () [0x00010] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacDebuggerObjectTypeView.cs:66 
  at MonoDevelop.Debugger.MacDebuggerObjectCellViewBase.set_ObjectValue (Foundation.NSObject value) [0x00085] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacDebuggerObjectCellViewBase.cs:79 
  at MonoDevelop.Debugger.MacObjectValueTreeViewDelegate.GetView (AppKit.NSOutlineView outlineView, AppKit.NSTableColumn tableColumn, Foundation.NSObject item) [0x000ca] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacObjectValueTreeViewDelegate.cs:73 
--- End of stack trace from previous location where exception was thrown ---

  at (wrapper managed-to-native) ObjCRuntime.Messaging.void_objc_msgSendSuper_IntPtr_IntPtr_UInt64(intptr,intptr,intptr,intptr,ulong)
  at AppKit.NSOutlineView.InsertItems (Foundation.NSIndexSet indexes, Foundation.NSObject parent, AppKit.NSTableViewAnimation animationOptions) [0x00099] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.16.1.24/src/Xamarin.Mac/AppKit/NSOutlineView.g.cs:407 
  at MonoDevelop.Debugger.MacObjectValueTreeViewDataSource.Append (System.Collections.Generic.IList`1[T] nodes) [0x000b8] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacObjectValueTreeViewDataSource.cs:249 
  at MonoDevelop.Debugger.MacObjectValueTreeView.MonoDevelop.Debugger.IObjectValueTreeView.Appended (System.Collections.Generic.IList`1[T] nodes) [0x00001] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacObjectValueTreeView.cs:501 
  at MonoDevelop.Debugger.ObjectValueTreeViewController+<>c__DisplayClass48_0.<AddValues>b__0 () [0x00001] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/ObjectValueTreeViewController.cs:266 
  at MonoDevelop.Core.Runtime.RunInMainThread (System.Action action) [0x00012] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/core/MonoDevelop.Core/MonoDevelop.Core/Runtime.cs:332 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.ArgumentNullException: Value cannot be null.
Parameter name: str
  at Foundation.NSAttributedString..ctor (System.String str, Foundation.NSDictionary attributes) [0x00098] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.16.1.24/src/Xamarin.Mac/Foundation/NSAttributedString.g.cs:298 
  at Foundation.NSAttributedString..ctor (System.String str, AppKit.NSFont font, AppKit.NSColor foregroundColor, AppKit.NSColor backgroundColor, AppKit.NSColor strokeColor, AppKit.NSColor underlineColor, AppKit.NSColor strikethroughColor, AppKit.NSUnderlineStyle underlineStyle, AppKit.NSUnderlineStyle strikethroughStyle, AppKit.NSParagraphStyle paragraphStyle, System.Single strokeWidth, AppKit.NSShadow shadow, Foundation.NSUrl link, System.Boolean superscript, AppKit.NSTextAttachment attachment, Foundation.NSLigatureType ligature, System.Single baselineOffset, System.Single kerningAdjustment, System.Single obliqueness, System.Single expansion, AppKit.NSCursor cursor, System.String toolTip, System.Int32 characterShape, AppKit.NSGlyphInfo glyphInfo, Foundation.NSArray writingDirection, System.Boolean markedClauseSegment, AppKit.NSTextLayoutOrientation verticalGlyphForm, AppKit.NSTextAlternatives textAlternatives, AppKit.NSSpellingState spellingState) [0x00000] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.16.1.24/src/Xamarin.Mac/Foundation/NSAttributedString.mac.cs:54 
  at MonoDevelop.Debugger.MacDebuggerObjectCellViewBase.GetAttributedString (System.String text) [0x0000e] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacDebuggerObjectCellViewBase.cs:130 
  at MonoDevelop.Debugger.MacDebuggerObjectTypeView.UpdateContents () [0x00010] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacDebuggerObjectTypeView.cs:66 
  at MonoDevelop.Debugger.MacDebuggerObjectCellViewBase.set_ObjectValue (Foundation.NSObject value) [0x00085] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacDebuggerObjectCellViewBase.cs:79 
  at MonoDevelop.Debugger.MacObjectValueTreeViewDelegate.GetView (AppKit.NSOutlineView outlineView, AppKit.NSTableColumn tableColumn, Foundation.NSObject item) [0x000ca] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacObjectValueTreeViewDelegate.cs:73 
--- End of stack trace from previous location where exception was thrown ---

  at (wrapper managed-to-native) ObjCRuntime.Messaging.void_objc_msgSendSuper_IntPtr_IntPtr_UInt64(intptr,intptr,intptr,intptr,ulong)
  at AppKit.NSOutlineView.InsertItems (Foundation.NSIndexSet indexes, Foundation.NSObject parent, AppKit.NSTableViewAnimation animationOptions) [0x00099] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.16.1.24/src/Xamarin.Mac/AppKit/NSOutlineView.g.cs:407 
  at MonoDevelop.Debugger.MacObjectValueTreeViewDataSource.Append (System.Collections.Generic.IList`1[T] nodes) [0x000b8] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacObjectValueTreeViewDataSource.cs:249 
  at MonoDevelop.Debugger.MacObjectValueTreeView.MonoDevelop.Debugger.IObjectValueTreeView.Appended (System.Collections.Generic.IList`1[T] nodes) [0x00001] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacObjectValueTreeView.cs:501 
  at MonoDevelop.Debugger.ObjectValueTreeViewController+<>c__DisplayClass48_0.<AddValues>b__0 () [0x00001] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/ObjectValueTreeViewController.cs:266 
  at MonoDevelop.Core.Runtime.RunInMainThread (System.Action action) [0x00012] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/core/MonoDevelop.Core/MonoDevelop.Core/Runtime.cs:332 <---

2019-09-26 10:23:50.075 MonoDevelop[56450:2194236] *** Assertion failure in -[NSTableRowData removeRowsAtIndexes:withRowAnimation:], /BuildRoot/Library/Caches/com.apple.xbs/Sources/AppKit/AppKit-1671.60.107/TableView.subproj/NSTableRowData.m:7444
FATAL ERROR [2019-09-26 10:23:50Z]: Unhandled ObjC Exception
MonoDevelop.MacIntegration.MacPlatformService+MarshalledObjCException: NSInternalInconsistencyException: removeRowsAtIndexes:withRowAnimation: can not happen while updating visible rows!
  at MonoDevelop.MacIntegration.MacPlatformService+MarshalledObjCException..ctor (Foundation.NSException exception) [0x00009] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MacPlatform/MacPlatform.cs:264 
  at MonoDevelop.MacIntegration.MacPlatformService.HandleUncaughtException (System.IntPtr exceptionPtr) [0x0001a] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MacPlatform/MacPlatform.cs:248 
  at ObjCRuntime.Messaging.void_objc_msgSendSuper_IntPtr_IntPtr_UInt64 (System.IntPtr , System.IntPtr , System.IntPtr , System.IntPtr , System.UInt64 ) [0x00000] in <9516c6ab0e7e466b9565e32a3dc558f9>:0 
  at AppKit.NSOutlineView.RemoveItems (Foundation.NSIndexSet indexes, Foundation.NSObject parent, AppKit.NSTableViewAnimation animationOptions) [0x00099] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.16.1.24/src/Xamarin.Mac/AppKit/NSOutlineView.g.cs:574 
  at MonoDevelop.Debugger.MacObjectValueTreeViewDataSource.Replace (MonoDevelop.Debugger.ObjectValueNode node, MonoDevelop.Debugger.ObjectValueNode[] replacementNodes) [0x000af] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacObjectValueTreeViewDataSource.cs:134 
  at MonoDevelop.Debugger.MacObjectValueTreeView.OnEvaluationCompleted (MonoDevelop.Debugger.ObjectValueNode node, MonoDevelop.Debugger.ObjectValueNode[] replacementNodes) [0x0000d] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacObjectValueTreeView.cs:444 
  at MonoDevelop.Debugger.MacObjectValueTreeView.LoadEvaluatedNode (MonoDevelop.Debugger.ObjectValueNode node, MonoDevelop.Debugger.ObjectValueNode[] replacementNodes) [0x00001] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/Mac/MacObjectValueTreeView.cs:451 
  at MonoDevelop.Debugger.ObjectValueTreeViewController+<>c__DisplayClass90_0.<OnEvaluationCompleted>b__0 () [0x00001] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/ObjectValue/ObjectValueTreeViewController.cs:836 
  at MonoDevelop.Core.Runtime+<>c.<RunInMainThread>b__43_0 (System.Object state) [0x00015] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/core/MonoDevelop.Core/MonoDevelop.Core/Runtime.cs:341 
  at MonoDevelop.Ide.DispatchService+GtkSynchronizationContext+TimeoutProxy.HandlerInternal (System.IntPtr data) [0x00016] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/DispatchService.cs:77 
  at Gtk.Application.gtk_main () [0x00000] in <426a4590353f426897d5a3e1eb51b9b8>:0 
  at Gtk.Application.Run () [0x00001] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08_2/external/bockbuild/builds/gtk-sharp-None/gtk/Application.cs:145 
  at MonoDevelop.Ide.IdeStartup.Run (MonoDevelop.Ide.MonoDevelopOptions options) [0x00512] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/IdeStartup.cs:230 
  at MonoDevelop.Ide.IdeStartup.Main (System.String[] args, MonoDevelop.Ide.Extensions.IdeCustomizer customizer) [0x000d4] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/IdeStartup.cs:736 
  at MonoDevelop.Startup.MonoDevelopMain.Main (System.String[] args) [0x00001] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/core/MonoDevelop.Startup/MonoDevelop.Startup/MonoDevelopMain.cs:40 
--- End of stack trace from previous location where exception was thrown ---

  at MonoDevelop.MacIntegration.MacPlatformService.HandleUncaughtException (System.IntPtr exceptionPtr) [0x0001a] in /Users/iain/Projects/xamarin/custom/master/monodevelop/main/src/addins/MacPlatform/MacPlatform.cs:248 
2019-09-26 10:23:50.479 MonoDevelop[56450:2194236] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'removeRowsAtIndexes:withRowAnimation: can not happen while updating visible rows!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff3ae732fd __exceptionPreprocess + 256
	1   libobjc.A.dylib                     0x00007fff65547a17 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff3ae8e016 +[NSException raise:format:arguments:] + 98
	3   Foundation                          0x00007fff3d12d791 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 194
	4   AppKit                              0x00007fff3861eaf1 -[NSTableRowData removeRowsAtIndexes:withRowAnimation:] + 350
	5   AppKit                              0x00007fff386023b7 -[NSTableView _doUpdatedWorkWithHandler:] + 74
	6   AppKit                              0x00007fff3861e845 -[NSTableView removeRowsAtIndexes:withAnimation:] + 139
	7   AppKit                              0x00007fff3879b44e -[NSOutlineView removeRowsAtIndexes:withAnimation:] + 41
	8   AppKit                              0x00007fff38ac81a8 -[NSOutlineView _removeItemsAtIndexes:inParentRowEntry:withAnimation:] + 1003
	9   AppKit                              0x00007fff386023b7 -[NSTableView _doUpdatedWorkWithHandler:] + 74
	10  AppKit                              0x00007fff38ac99fd -[NSOutlineView removeItemsAtIndexes:inParent:withAnimation:] + 252
	11  libxammac.dylib                     0x00000001066eedac xamarin_dyn_objc_msgSendSuper + 220
	12  ???                                 0x0000000140a7a607 0x0 + 5379696135
)
```

iainx commented:
The locals pad will be accessible once #8807 is merged

jstedfast commented:
@DavidKarlas 

> 1. When number of lines in Locals pad goes beyond size of Locals pad scrolling doesn't work:

Can't replicate this issue...

> 2. Resizing of tooltip doesn't seem to ideal: In this case its too big:

Fixed?

> Also there is some extra space after Pin icon that seems unnessesary.

Fixed.

> Preview icon, doesn't seem to work:

Fixed.

jstedfast commented:
@iainx the crash should be fixed now

jstedfast commented:
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/782188/

jstedfast commented:
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/934130/
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/957702/
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/888712/
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/901729/
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/782188/

DavidKarlas commented:
I'm able to reproduce problem with scrolling inside Locals only for CI built aka. downloaded from CI .dmg builds.
I can't reproduce for locally built VSmac...
There is also difference in header background color:
Local build:
<img width="568" alt="image" src="https://user-images.githubusercontent.com/774791/65812461-83f31c00-e1c7-11e9-9004-08bd299cd683.png">
Dmg from CI:
<img width="573" alt="image" src="https://user-images.githubusercontent.com/774791/65812469-92d9ce80-e1c7-11e9-9cf2-b2bc8cdc2598.png">

jstedfast commented:
When you are running the DMG from CI, I think the reason there is a slight difference is because one is Xamarin.Mac (local build) and one is Gtk (CI build).

You can specify which one to use by going into Debugger preferences and selecting which one to use.

DavidKarlas commented:
I have new enabled... can you try installing .dmg from CI?

vancura commented:
I just figured out, could you please use `~sel` icons when the row is selected?

<img width="38" alt="image" src="https://user-images.githubusercontent.com/4982/65871534-e80c1080-e37e-11e9-8b6e-a19e1d31d380.png">

jstedfast commented:
There are a number of issues still remaining in this PR and this comment is here to try and track them (thus it will be updated as new issues are found or issues become resolved).

These are the known issues remaining:

- [ ] In local builds of this PR, the Locals & Watch pads are scrollable no matter how many nodes in the list are expanded, but in the CI builds, this is not the case. When using a CI build, expanding nodes seems to trick the `NSScrollView` containing the `MacObjectValueTreeView` (a subclass of `NSOutlineView`) into thinking that it has far more visible real estate than it actually has and therefor it does not allow you to scroll. This is just a theory, but it makes sense, I think? Perhaps an issue with GtkNSHostView?
- [ ] Adding a new expression in the Watch Pad only semi-sorta works. For some reason, when adding an expression that would require an expander (e.g. if you add the expression `buffer` which is defined as `new byte[123]`), the expression string is NOT displayed in the new row that gets added (but the expander, icon, and preview eyeball icon thingy *do* get shown). Based on what I can figure out so far, the expression string *is* being set on the `TextField`, but it's not rendering... is it just oddly sized to be 0px wide or something??
- [ ] In the Watch Pad (and maybe Locals Pad too? unconfirmed), sometimes when an expression that has a visualizer icon in the Value column is pushed off-screen (due to an expression above it being expanded, for example), when you collapse said expanded node allowing the other expression back into the viewport, the visualizer icon is gone. Once it is gone, it seems to never come back.
- [ ] While @sgmunn was playing around with the Locals Pad, he discovered that expanding the "Non-Public Members" row, it would insert X-number of completely blank rows into the treeview (presumably where X matches the number of non-public members). See the screenshot below: <img width="710" alt="Screen Shot 2019-10-03 at 11 18 02 AM" src="https://user-images.githubusercontent.com/338984/66142899-45e56600-e5d4-11e9-8dbf-e2f7998aa148.png">
- [ ] Greg has also been able to confuse the Locals Pad into display the preview window in the wrong location: <img width="512" alt="Screen Shot 2019-10-03 at 11 18 20 AM" src="https://user-images.githubusercontent.com/338984/66142969-69a8ac00-e5d4-11e9-980a-6e6e63c96079.png"> When Greg got encountered this bug, he had first resized the Locals Pad to be much larger and then back to a smaller size, so this *may* be required step to reproduce this bug (unconfirmed).

sandyarmstrong commented:
Just started playing with this. Only feedback so far is to disable animations. Looks like the editor extension to do this isn't public yet, but it looks like this:

```csharp
        static readonly NSDictionary disableAnimations = new NSDictionary(
            "actions", NSNull.Null,
            "contents", NSNull.Null,
            "hidden", NSNull.Null,
            "onLayout", NSNull.Null,
            "onOrderIn", NSNull.Null,
            "onOrderOut", NSNull.Null,
            "position", NSNull.Null,
            "sublayers", NSNull.Null,
            "transform", NSNull.Null,
            "bounds", NSNull.Null);

        public static void DisableImplicitAnimations(this CALayer layer)
            => layer.Actions = disableAnimations;
```

sgmunn commented:
so, scrolling improved but rendering glitches galore - 

![Screen Recording 2019-10-04 at 08 45 AM](https://user-images.githubusercontent.com/723721/66208606-76380d80-e683-11e9-9693-f325611ef4cc.gif)

sgmunn commented:
@jstedfast a couple of nodes are failing to load

https://cl.ly/906d9abf8760

sgmunn commented:
@sandyarmstrong do we disable the animations on the layer of the NSOutlineView itself, the scrollview or somewhere else?

sgmunn commented:
not critical for the first preview, but the loading logic could use some tweaking so we don't have visual ghosting - watch the following frame by frame as the node expands.

https://cl.ly/f62f911ecd8c

sgmunn commented:
rendering issues when expanding and collapsing. I had an example where there was text overwriting each other but failed to capture it on a second go. this clip still shows some not good rendering - probably the tree not being updated properly.

![Screen Recording 2019-10-07 at 09 56 AM](https://user-images.githubusercontent.com/723721/66318242-f4491e00-e8e8-11e9-9bdc-235061176215.gif)

sgmunn commented:
I am inclined to think that the rendering issue on the tooltip is somehow related to the editor drawing surface but I cannot think of a reason how. I've looked at various ideas like we're adding additional fields to the view, constraints, and loading wrong and I don't see anything amiss. Compare the two gifs below for expanding the same value and you can see the issue with the tooltip. The other reason I think it might be related to the editor surface is that even the macOS expanders are not being drawn correctly and ghost when expanded - not just the things that we draw.

![Screen Recording 2019-10-07 at 06 40 PM](https://user-images.githubusercontent.com/723721/66354380-9ccfa000-e932-11e9-81e4-ba525ace36b4.gif)

![Screen Recording 2019-10-07 at 06 42 PM](https://user-images.githubusercontent.com/723721/66354383-9fca9080-e932-11e9-9859-a2f3fb160692.gif)

sgmunn commented:
@sandyarmstrong does any of the tooltip rendering issues make sense?

sandyarmstrong commented:
@sgmunn I'll look in the morning. I thought I saw all those same issues in the pad when I was testing earlier, though.

DavidKarlas commented:
Just passing by... Are animations disabled or enabled?

jstedfast commented:
They are disabled

sgmunn commented:
@sandyarmstrong can you explain what the animation disabling does and why we need it? To be honest we didn't notice much of a difference so not sure what we're looking at here. If you can review that would be great.

sgmunn commented:
redeploying unit tests since they seemed to timeout with fetching the app bundle

sandyarmstrong commented:
@sgmunn which animations are we talking about? I disabled animations on the NSPopover because they slow down the user experience, and in general we don't want these kind of animations in editor adornments unless UX Board is asking for them specifically.

sgmunn commented:
@sandyarmstrong I'm referring to this comment - https://github.com/mono/monodevelop/pull/8461#issuecomment-538148320

I had assumed you meant animations on the treeviw so added what I thought was what you meant. however, it was that that also lead to odd header transparency issues until after I added the same code to the scroll view. In the end, neither Jeff or I could tell (visually) what the code was doing.

sgmunn commented:
DDRits and compat check failures are not unique to this branch

sandyarmstrong commented:
@sgmunn ah, ok. well, I meant animations everywhere. The code I shared was to stop implicit Cocoa animations. At least, that's what we used to stop built-in animations in the editor whenever the user added a new line, for example.

For the popover, it was as simple as setting a single property. For other things it may be more complex. I don't think animations should be a blocker here in any way, though. You should do whatever you need to do to get this working. :)

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.PreviewVisualizers/PreviewWindowManager.cs:
```diff
@@ -47,10 +47,16 @@ public static void Show (ObjectValue val, Control widget, Gdk.Rectangle previewB
 			wnd = new PreviewVisualizerWindow (val, widget);
 			IdeApp.CommandService.RegisterTopWindow (wnd);
 			wnd.ShowPopup (widget, previewButtonArea, PopupPosition.Left);
+			wnd.FocusOutEvent += HandleFocusOutEvent;
```
this handler needs to be removed.

jstedfast replied:
I have a patch for that now, but assume I should not commit/push until this has been merged.

manish commented:
A temporary fix for DDRIT is to remove `Sanity` category here https://github.com/xamarin/md-addins/blob/master/Tests/Microsoft.VSForMac.Tests/Resources/DataSources/DebuggingDataFactory.cs#L108
